### PR TITLE
Rinkeby deployment

### PR DESCRIPTION
**Pull request by jamesmorgan**
* [x] Rinkeby deployment for KO
* [x] clean up code
* [x] add token marketplace logic

-----

* [x] Contracts - knownorigin/known-origin-contracts#3
* [x] Subgraph - knownorigin/known-origin-subgraph#6
* [x] Frontend - knownorigin/known-origin-web3-marketplace#209
* [x] API - https://github.com/knownorigin/known-origin-serverless-api/pull/45

### CODE REVIEW

File: build/schema.graphql [hunk 1]
```diff
@@ -1,24 +1,51 @@
...
+    name: String!
+    token: Token!
+    edition: Edition!
+    buyer: Collector
+    bidder: Collector
+    currentOwner: Collector
+    ethValue: BigDecimal!
+    timestamp: BigInt!
+    transactionHash: Bytes!
+}
```
Comment (andygray) [inline build/schema.graphql:18]: 👍

File: build/schema.graphql [hunk 2]
```diff
@@ -39,27 +67,35 @@ type Day @entity {
...
+    transferCount: BigInt!
+    editionNumber: BigInt!
+    tokenURI: String!
+    metadata: MetaData!
+    birthTimestamp: BigInt!
+    primaryValueInEth: BigDecimal!
+    lastTransferTimestamp: BigInt!
+
+    currentOwner: Collector!            # the current owner
+    allOwners: [Collector]!             # list (set) of all unique owners
```
Comment (andygray) [inline build/schema.graphql:87]: should they all be [thing!]! ?

File: src/mappings/token-marketplace-mapping.ts [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    */
+    let contract = getKnownOriginForAddress(event.address)
+
+    let tokenEntity = loadOrCreateToken(event.params._tokenId, contract, event.block)
+    tokenEntity.openOffer = null
+    tokenEntity.currentTopBiider = null
+    tokenEntity.save();
+
+    clearTokenOffer(event.block, contract, event.params._tokenId)
+}
```
Comment (andygray) [inline src/mappings/token-marketplace-mapping.ts:79]: nice...

File: src/mappings/token-marketplace-mapping.ts [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    */
+    let contract = getKnownOriginForAddress(event.address)
+
+    createBidWithdrawnEvent(event)
+    clearTokenOffer(event.block, contract, event.params._tokenId)
+
+    let tokenEntity = loadOrCreateToken(event.params._tokenId, contract, event.block)
+    tokenEntity.openOffer = null
+    tokenEntity.currentTopBiider = null
+    tokenEntity.save();
```
Comment (andygray) [inline src/mappings/token-marketplace-mapping.ts:185]: this code repeated a few times - might benefit from a function like `clear`

`currentTopBiider` spelling mistake

File: src/services/MetaData.service.ts [hunk 5]
```diff
@@ -13,23 +15,38 @@ export function constructMetaData(tokenURI: string): MetaData | null {
...
         if (data !== null) {
             let jsonData: JSONValue = json.fromBytes(data as Bytes)
-            metaData.name = jsonData.toObject().get('name').toString()
-            metaData.description = jsonData.toObject().get('description').toString()
-            metaData.image = jsonData.toObject().get('image').toString()
 
+            if (isObject(jsonData) && jsonData.toObject().isSet('name')) {
+                metaData.name = jsonData.toObject().get('name').toString()
+            } else {
+                metaData.name = "NOT SET"
```
Comment (andygray) [inline src/services/MetaData.service.ts:22]: thought: would `null` be better than "NOT SET"?

File: src/services/TokenEvent.factory.ts [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+
+export function createTokenPrimaryPurchaseEvent(event: Purchase): TokenEvent {
+    let timestamp = event.block.timestamp;
+
+    let tokenEventId = "Purchase-"
+        .concat(event.params._tokenId.toString())
+        .concat("-")
+        .concat(event.params._buyer.toHexString())
+        .concat("-")
+        .concat(timestamp.toString());
```
Comment (andygray) [inline src/services/TokenEvent.factory.ts:26]: now that's what a I call a key...

File: src/services/MetaData.service.ts (hunk 5 above)
Comment (jamesmorgan) [reply src/services/MetaData.service.ts:22]: Maybe, it is a non nullable field though - I was going to set it to `ERROR` instead - thoughts?

File: src/mappings/token-marketplace-mapping.ts (hunk 4 above)
Comment (jamesmorgan) [reply src/mappings/token-marketplace-mapping.ts:185]: fixed

File: build/schema.graphql (hunk 2 above)
Comment (jamesmorgan) [reply build/schema.graphql:87]: fixed

**Pull request merged**